k3k-kubelet/provider/provider.go

```diff
@@ -249,28 +255,44 @@ func (p *Provider) CreatePod(ctx context.Context, pod *corev1.Pod) error {
 // if one/more volumes couldn't be transformed
 func (p *Provider) transformVolumes(ctx context.Context, podNamespace string, volumes []corev1.Volume) error {
 	for _, volume := range volumes {
+		var optional bool
+		if strings.HasPrefix(volume.Name, kubeAPIAccessPrefix) {
+			continue
+		}
 		// note: this needs to handle downward api volumes as well, but more thought is needed on how to do that
 		if volume.ConfigMap != nil {
-			if err := p.syncConfigmap(ctx, podNamespace, volume.ConfigMap.Name); err != nil {
+			if volume.ConfigMap.Optional != nil {
+				optional = *volume.ConfigMap.Optional
+			}
+			if err := p.syncConfigmap(ctx, podNamespace, volume.ConfigMap.Name, optional); err != nil {
 				return fmt.Errorf("unable to sync configmap volume %s: %w", volume.Name, err)
 			}
 			volume.ConfigMap.Name = p.Translater.TranslateName(podNamespace, volume.ConfigMap.Name)
 		} else if volume.Secret != nil {
-			if err := p.syncSecret(ctx, podNamespace, volume.Secret.SecretName); err != nil {
+			if volume.Secret.Optional != nil {
+				optional = *volume.Secret.Optional
+			}
+			if err := p.syncSecret(ctx, podNamespace, volume.Secret.SecretName, optional); err != nil {
```

Comment:
Imho the code will look more readable changing the structure to have less branches, and passing the struct to the `syncXXX` funcs. Something like:

```go
if volume.ConfigMap != nil {
	if err := p.syncConfigmap(ctx, podNamespace, volume.ConfigMap; err != nil {
		return fmt.Errorf("unable to sync configmap volume %s: %w", volume.Name, err)
	}
	volume.ConfigMap.Name = p.Translater.TranslateName(podNamespace, volume.ConfigMap.Name)
	continue
}

if volume.Secret != nil {
	// same
}
```

What do you think?

Reply:
that makes sense, fixing and testing now

Reply:
hmmm, I think why Michael didnt want to do that in the first place, simply because there are 2 types of volumes , there is simple configmap and projection configmap, which are two different types, thats why he only sent the names and not the whole object, that said Michael wanted to make the syncer more generic anyway, so I think we can merge for now and for future iteration we can conver the whole syncer to a more generic functions to accept any type of object

Reply:
Yes, I've just realized that. :/

Kinda annoying but I agree, thanks anyway! :+1: